### PR TITLE
Develop multi discrete

### PR DESCRIPTION
**Pull request by vincentpierre**
# Abstract 
 
[Recent advancements](https://arxiv.org/abs/1807.01281) have been made in Discrete Control by using multiple discrete actions concurently. This branching of actions could allow us to solve harder problems such as locomotion and puzzle solving in discrete control.

# Background

Discrete Control only allows the use of one action at a time. In the WallJump Environment's current version, the agent can either move or jump __not__ the two at the same time. We aim to change that.

# Goals

 - Implement Multi Discrete control
 - Demonstrate efficiency with a locomotion or puzzle game
 - Maintain the same level of performance

# Non-Goals

 - Retrain Environments
 - Multi Continuous control

# Plan

 - [x] Make the VectorActionSize in the protobuff definition of BrainParameters an array of int instead of an int
 - [x] Update the Brain Parameters for the Example Environments
 - [x] Train Crawler with Discrete Control

# Acceptance Criteria

 - No loss in performance for one Branch in the current discrete environment
 - Enables to solve simple locomotion with Discrete Control

# Other considerations

[Action Masking](https://github.com/Unity-Technologies/ml-agents/pull/931) is dependent on this change
Will need to retrain the models before release

### CODE REVIEW

**Comment by awjuliani** [comment]
A couple high-level points:

* There should be accompanying documentation changes.
* Since you changed the meaning of all of the discrete actions in the example environments, the models will need to be re-trained.

Will provide a more in-depth code review soon.

**Comment by awjuliani** [comment]
Beginning review. In the UI under "Branches" there is currently a "Size" and multiple "Element X" values. Would it be possible to rename these to "Number of Branches" and "Branch Size?" 

screen shot 2018-07-31 at 11 49 30 am

**Comment by vincentpierre** [comment]
Not without writing custom Unity UI code which no one should ever be forced to do. :( I will try something out...

File: python/unitytrainers/ppo/trainer.py [hunk 1]
```diff
@@ -165,6 +165,7 @@ def take_action(self, all_brain_info: AllBrainInfo):
         :return: a tuple containing action, memories, values and an object
         to be passed to add experiences
         """
+        action_sizes = [3, 3, 3, 2]
```
[the remark this answers is not in the document]
Comment (vincentpierre) [reply python/unitytrainers/ppo/trainer.py]: Dead code

File: python/unitytrainers/models.py [hunk 2]
```diff
@@ -287,9 +249,66 @@ def create_cc_actor_critic(self, h_size, num_layers):
...
+        self.all_probs = tf.concat([tf.nn.softmax(branch) for branch in policy_branches], axis=1, name="action_probs")
+
+        output = tf.concat([tf.multinomial(branch, 1) for branch in policy_branches], axis=1)
+
+        self.output = tf.identity(output, name="action")
+
+        value = tf.layers.dense(hidden, 1, activation=None)
+        self.value = tf.identity(value, name="value_estimate")
+
+        self.entropy = - (sum([tf.reduce_sum(tf.nn.softmax(branch) * tf.log(tf.nn.softmax(branch) + 1e-10), axis=1)
```
[the remark this answers is not in the document]
Comment (vincentpierre) [reply python/unitytrainers/models.py]: We are doing the average of the entropy of each branch

**Comment by awjuliani** [comment]
The `BananaCollector` environment should be converted to Multi-discrete as well.

**Comment by vincentpierre** [comment]
Not sure if we should modify more environments in the scope of this PR

File: python/unitytrainers/models.py [hunk 3]
```diff
@@ -287,9 +249,66 @@ def create_cc_actor_critic(self, h_size, num_layers):
...
+            self.prev_action = tf.placeholder(shape=[None, len(self.a_size)], dtype=tf.int32, name='prev_action')
+            prev_action_oh = tf.concat([
+                tf.one_hot(self.prev_action[:, i], self.a_size[i]) for i in range(len(self.a_size))], axis=1)
+            hidden = tf.concat([hidden, prev_action_oh], axis=1)
+
+            self.memory_in = tf.placeholder(shape=[None, self.m_size], dtype=tf.float32, name='recurrent_in')
+            hidden, memory_out = self.create_recurrent_encoder(hidden, self.memory_in, self.sequence_length)
+            self.memory_out = tf.identity(memory_out, name='recurrent_out')
+
+        policy_branches = []
```
[the remark this answers is not in the document]
Comment (vincentpierre) [reply python/unitytrainers/models.py]: There is not too much that is shared. It is just the branching of the policy. I would keep them separated for now.

File: unity-environment/Assets/ML-Agents/Editor/BrainEditor.cs [hunk 4]
```diff
@@ -80,13 +75,49 @@ public override void OnInspectorGUI()
...
+                        "Space Size", "Length of continuous action vector."));
+                    
+                }
+                else
+                {
+                    // Discrete case :
+                    SerializedProperty bpVectorActionSize =
+                        serializedBrain.FindProperty("brainParameters.vectorActionSize");
+                    bpVectorActionSize.arraySize = EditorGUILayout.IntField(
+                        "Nb Branches", bpVectorActionSize.arraySize);
```
[the remark this answers is not in the document]
Comment (xiaomaogy) [reply unity-environment/Assets/ML-Agents/Editor/BrainEditor.cs]: Also another side note, is there a way for us to have a toggle so that our users will still have the our previous display, and if they check the toggle "enable multi discrete action", they will have this new display. 

The reason I bring this up is that having "Nb Branches" under "Space Type" (our new UI) is unintuitive compare to having "Space Size" under "Space Type" (our previous UI). And I saw so many people give up trying or continue to use ml-agents because they have backward compatibility issues (I lurk in some of the ml-agents user group chat and watch them discuss about our product).

[the remark this answers is not in the document]
Comment (vincentpierre) [reply unity-environment/Assets/ML-Agents/Editor/BrainEditor.cs]: Unfortunately, the gaol of this PR is to replace entirely the way we do Discrete Control, so we will have back-compatibility issue.

**Comment by vincentpierre** [comment]
Crawler now learning with discrete control. Review ready.

**Comment by xiaomaogy** [comment]
On the Match3 problem, this multi-discrete branch is working better than the previous multi-discrete version and the non-multi-discrete version.
